src/lib.rs

```diff
@@ -1311,6 +1315,123 @@ where
     Some(Ratio::new(n1, d1))
 }
 
+#[cfg(all(feature = "bigint", feature = "std"))]
+impl<T: Clone + Integer + Signed + ToPrimitive + ToBigInt> Ratio<T> {
+    /// Converts the ratio to an `f64`.
+    ///
+    /// The value is guaranteed to be the closest `f64` approximating the ratio. This always
+    /// returns a value, unless the conversion to BigInt for the numerator or denominator fails.
+    ///
+    /// ```
+    /// use num_rational::Rational64;
+    /// assert_eq!(0.5f64, Rational64::new(1, 2).to_f64().unwrap());
+    /// ```
+    pub fn to_f64(&self) -> Option<f64> {
+        assert_eq!(
+            std::f64::RADIX,
+            2,
+            "only floating point implementations with radix 2 are supported"
+        );
+
+        // Upper bound to the range of exactly-representable ints in an f64.
+        const MAX_EXACT_INT: u64 = 1u64 << std::f64::MANTISSA_DIGITS;
+
+        let numer: T = self.numer().abs();
+        let denom: T = self.denom().abs();
+        let sign: T = self.numer().signum() * self.denom().signum();
+        let flo_sign = if sign.is_negative() { -1.0 } else { 1.0 };
+
+        if numer.is_zero() {
+            return Some(0.0 * flo_sign);
+        }
+
+        // Fast track: both sides can losslessly be converted to f64s. In this case, letting the
+        // FPU do the job is faster and easier. In any other case, converting to f64s may lead
+        // to an inexact result: https://stackoverflow.com/questions/56641441/
+        if numer
```

Comment:
nitpick: I think it would be fewer lines / more idiomatic to match on (numer.to_u64(), denom.to_u64()) here

Reply:
Agreed, I used this `and_then` business to avoid the second conversion if the first fails, but the Rust compiler may figure this out, and there's enough BigInt arithmetic going on in this method that it probably won't make a big performance difference.

Do you have an opinion on whether I should replace `Some(<convert to float>.unwrap())` in the lines below with just the conversion code? I think the conversion is guaranteed to succeed (but not completely sure—the `ToPrimitive` trait documentation doesn't explicitly say what the allowed failure cases are, and in particular whether it's allowed to return `None` when the value fits). In any case, I think I should either allow the conversion to fail and carry on with the BigInt path, or assume it never fails, in which case I can just return the result of the `and_then` expression.